README
NAME
    XUL::App - Nifty XUL apps in a XUL::App

VERSION
    This document describes XUL::App 0.07 released on July 11, 2008.

SYNOPSIS
        $ xulapp app --name YSearchAll

        $ cd YSearchAll

        $ xulapp view --name Overlay --type overlay

        # Edit lib/YSearchAll/App.pm to add the following lines:
        package YSearchAll::App;
        our $VERSION = '0.07';

        use XUL::App::Schema;
        use XUL::App schema {
        +     xulfile 'overlay.xul' =>
        +         generated from 'YSearchAll::View::Overlay',
        +         includes qw( xulapp/jquery.js overlay.js ),
        +         overlays 'chrome://browser/content/browser.xul';
        +
            xpifile 'ysearchall.xpi' =>
                name is 'YSearchAll',
                id is 'ysearchall@agentz.agentz-office',
                version is '0.0.1',
                targets {
                    Firefox => ['2.0' => '3.0.*'],  # FIXME
                    Mozilla => ['1.5' => '1.8'],  # FIXME
                },
                creator is 'The YSearchAll development team',
                developers are ['agentz'],
                contributors are [],
                homepageURL is 'http://searchall.agentz.org', # FIXME
                iconURL is '',  # like 'chrome://helloworld/content/logo.png'
                updateURL is ''; # This should not set for AMO extensions.
        };
        1;

        # Add and edit js/ysearchall.js manually
        $ xulapp overlay -p dev11
        $ xulapp bundle .  # generate XPI ready for deployment

DESCRIPTION
    XUL::App is a nifty Firefox extension development framework based on
    Perl. It has a lot of parallels with Jifty. In particular, this
    framework allows you to build real-world modern Firefox extensions using
    Perl. But the resulting XPI installation file is completely portable and
    contains *0* Perl.

    This framework has the following highlights:

    *   Jifty love on the Firefox extension development land.

    *   Building a realworld Firefox extension can be much easier than
        GreaseMonkey hacks.

    *   Use 100% Perl to specify XUL layout via Template::Declare, no XML
        pain.

    *   Automatic extension registration/unregistration for easy debugging.
        And no more frustration from Firefox's XUL/ext cache. (XUL::App will
        ensure the developer version defeats Firefox's cache.)

    *   Transparent "jar"-based XPI bundling, no chrome.manifest,
        install.rdf, and zip panic. XUL::App generates all of them for you
        according to your (declarative) Perl code.

    *   The resulting XPI file contains 0% Perl and could run everywhere
        (Win32, Mac, Linux, and etc). It's a compiler-style framework.

    *   I18N support via the Locale::Maketext::Lexicon module (same as
        Jifty, but actually emulated by XUL's native I18N mechanism).

    Currently the module is still in alpha stage and we're in severe lack of
    documentation and tests. (Although a real-world fully-fledged Firefox
    extension named SearchAll
    (<https://addons.mozilla.org/en-US/firefox/addon/5712>) is already
    developed atop it.

    You can get a lot of information from the slides that I used in the
    talk:

    <http://agentzh.org/misc/slides/xulapp/xulapp.xul> (a JS-enabled Firefox
    is required to view these slides)

    If you're having problems in displaying the slides given above, please
    try out the PowerPoint (PPT) version below:

    <http://agentzh.org/misc/slides/xulapp.ppt>

    or the PDF version:

    <http://agentzh.org/misc/slides/xulapp.pdf>

CAVEATS
    *   Because XUL::App wites to your ~/.mozilla/firefox directory
        directly, it's always recommended to backup that directory to
        somewhere else to prevent unexpected corruption occurring to your
        Firefox profiles.

    *   After applying Firefox updates or using a new Firefox profile, it's
        required to restart Firefox one more time to load your extensions.
        So please don't fire bug report regarding that.

    *   Only Firefox 2.0.0.* and Firefox 3.0 on Linux has been tested
        against this framework. So you're warned when you're going to do
        XUL::App hacking on other platforms like Windows.

        (BTW, I'm very willing to apply patches for other platforms.)

SOURCE CONTROL
    You can always get the latest source of XUL::App from the following SVN
    repository:

    <http://svn.openfoundry.org/xulapp/trunk/>

    There's very little POD docs, please read README and the slides for
    details.

    I really need help in improving this module's docs, tests, and
    implementation. If you find this thing useful and feel like contributing
    to it, please write to me and get a commit bit! ;)

SAMPLES
    *   SearchAll is a real-world Firefox extension that is built upon
        XUL::App, which can serve as a big demo for the usage of the
        framework:

        <http://svn.openfoundry.org/searchall/trunk/>

        And it's already on the mozilla official site AMO:

        <https://addons.mozilla.org/en-US/firefox/addon/5712>

    *   ExportReader: A Firefox extension to dump entries from Google Reader
        to JSON:

        <http://svn.openfoundry.org/xulapp/trunk/demo/ExportReader/>

        This extension only costed me about 10 lines of Perl and 20 lines of
        JavaScript, in addition to a few shell commands. The process is easy
        and enjoyable.

        Because it's mostly for personal use, The usage of this addon
        deserves some explanation though:

        1.  Click the XPI file from your Firefox browser:

            <http://svn.openfoundry.org/xulapp/trunk/demo/ExportReader/expor
            treader.xpi>

            After installation, please remember to restart to browser.

        2.  Enter the following chrome URL in your Firefox's address bar:

            <chrome://exportreader/content/exportreader.xul>

            Then you will see the main UI of this extension.

        3.  Login to your Google Reader and then click the site on the
            left-hand-side menu whose entries are what you want to export,
            say, "chromatic's Journal". Ensure that the "Expanded View" is
            used in the right-hand-side on the Google Reader page.

        4.  Click the "Extract!" button on the right-upper corner. Then
            you'll see an alert dialog saying how many entries were found.
            Click "OK" then the corresponding JSON string will be in the
            right-most textbox which is ready for copy-and-paste.

            Note that, by defaul, Google Reader lazily loads just the top 5
            entries or so. So in order to export all the entries in the
            subscribed site, say, "chromatic's Journal", you'll have to
            scroll down the Reader's Expanded View and force it to retrieve
            more entries.

    *   A helloworld sample extension is given in my XUL::App talk's slides:

        <http://agentzh.org/misc/slides/xulapp.pdf>

INSTALLATION
        perl Makefile.PL
        make
        sudo make install

BUGS
    Sadly XUL::App does not run on Win32 yet. I've only tested it on Ubuntu
    Linux so far. If you have any problems or would love to help, please let
    me know ;)

AUTHOR
    Agent Zhang <agentzh@yahoo.cn>

COPYRIGHT AND LICENSE
    Copyright 2007, 2008 by Yahoo! China EEEE Works, Alibaba Inc.
    (<http://www.eeeeworks.org>)

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself, either Artistic or GPL.

SEE ALSO
    xulapp, Locale::Maketext::Lexicon, Template::Declare, Jifty.



Changes
Change Log for XUL::App

0.07 Jul 11, 2008
* Fixed a syntax error in JSON.js regarding Firefox 3.

0.06 Jun 24, 2008
* Fixed various typos in the POD.

0.05 Jun 24, 2008
* Fixed a typo in the boiler-plate for XXX::App where XXX is the user app's name.
* Added a display_name attribute to XUL::App::Schema
* Added a new demo YahooFavorites (http://help.cn.yahoo.com/answerpage.html?product=myweb).
* Added a new demo HelloWorld.
* Added a new script xml2td for converting XML files to Perl code using Template::Declare.

0.04 Jun 23, 2008
* Updated the POD a bit.

0.03 Jun 23, 2008
* Added a link to the ExportReader sample Firefox extension's XPI to the POD.

0.02 Jun 23, 2008
* added missing dependencies to Makefile.PL.

0.01 Jun 22, 2008
* initial CPAN release.



Makefile.PL
use strict;
use lib '.';
use inc::Module::Install;

name                ('XUL-App');
license             ('perl');
author              ('Agent Zhang <agentzh@yahoo.cn>');
perl_version        ('5.006001');
all_from            ('lib/XUL/App.pm');

requires            ('Object::Declare' => '0.22');
requires            ('Template::Declare' => '0.28');
requires            ('Template::Declare::Tags');
requires            ('File::Slurp');
requires            ('Class::Accessor::Fast' => '0.30');
requires            ('Class::Data::Inheritable' => '0.06');
requires            ('File::ShareDir' => '0.04');
requires            ('File::Copy::Recursive' => '0.35');
requires            ('File::Basename');
requires            ('Locale::Maketext::Lexicon');
requires            ('Locale::Maketext');
requires            ('Text::Diff');
requires            ('File::Find');
requires            ('Sys::Hostname');

no_index( directory => qw< t inc doc share demo > );

install_script      ('script/xulapp');
install_share;

auto_install();
WriteAll();



share/js/JSON.js
var JSON = (function () {
    var m = {
            '\b': '\\b',
            '\t': '\\t',
            '\n': '\\n',
            '\f': '\\f',
            '\r': '\\r',
            '"' : '\\"',
            '\\': '\\\\'
        },
        s = {
            'boolean': function (x) {
                return String(x);
            },
            number: function (x) {
                return isFinite(x) ? String(x) : 'null';
            },
            string: function (x) {
                if (/["\\\x00-\x1f]/.test(x)) {
                    x = x.replace(/([\x00-\x1f\\"])/g, function(a, b) {
                        var c = m[b];
                        if (c) {
                            return c;
                        }
                        c = b.charCodeAt();
                        return '\\u00' +
                            Math.floor(c / 16).toString(16) +
                            (c % 16).toString(16);
                    });
                }
                return '"' + x + '"';
            },
            object: function (x) {
                if (x) {
                    var a = [], b, f, i, l, v;
                    if (x instanceof Array) {
                        a[0] = '[';
                        l = x.length;
                        for (i = 0; i < l; i += 1) {
                            v = x[i];
                            f = s[typeof v];
                            if (f) {
                                v = f(v);
                                if (typeof v == 'string') {
                                    if (b) {
                                        a[a.length] = ',';
                                    }
                                    a[a.length] = v;
                                    b = true;
                                }
                            }
                        }
                        a[a.length] = ']';
                    } else if (x instanceof Object) {
                        a[0] = '{';
                        for (i in x) {
                            v = x[i];
                            f = s[typeof v];
                            if (f) {
                                v = f(v);
                                if (typeof v == 'string') {
                                    if (b) {
                                        a[a.length] = ',';
                                    }
                                    a.push(s.string(i), ':', v);
                                    b = true;
                                }
                            }
                        }
                        a[a.length] = '}';
                    } else {
                        return;
                    }
                    return a.join('');
                }
                return 'null';
            }
        };
    return {
        copyright: '(c)2005 JSON.org',
        license: 'http://www.crockford.com/JSON/license.html',
        stringify: function (v) {
            var f = s[typeof v];
            if (f) {
                v = f(v);
                if (typeof v == 'string') {
                    return v;
                }
            }
            return null;
        },
        parse: function (text) {
            try {
                return !(/[^,:{}\[\]0-9.\-+Eaeflnr-u \n\r\t]/.test(
                        text.replace(/"(\\.|[^"\\])*"/g, ''))) &&
                    eval('(' + text + ')');
            } catch (e) {
                return false;
            }
        }
    };
})();



lib/XUL/App/XPIFile.pm
package XUL::App::XPIFile;

use strict;
use warnings;
use XUL::App::View::Install;
use File::Slurp;

use base 'Class::Accessor::Fast';
__PACKAGE__->mk_accessors(qw{
    id name description version targets
    creator contributors developers
    homepageURL updateURL iconURL
});

sub new {
    my $proto = shift;
    my $self = $proto->SUPER::new(@_);
    if ($self->id) {
        $XUL::App::ID = $self->id;
        if ($self->name =~ /\.xpi$/i) {
            die "You have to define a name for your plugin";
        }
        $XUL::App::APP_NAME = lc($self->name);
    }
    return $self;
}

sub go {
    my ($self, $file) = @_;
    # generate tmp/install.rdf here:
    Template::Declare->init( roots => ['XUL::App::View::Install'] );
    my $rdf = Template::Declare->show('main', $self);
    write_file('tmp/install.rdf', $rdf);
}

1;



lib/XUL/App/View/Install.pm
package XUL::App::View::Install;

use strict;
use warnings;
#use Smart::Comments;

use base 'Template::Declare';
use Template::Declare::Tags
        'RDF::EM' => { namespace => 'em' }, 'RDF';

our %UUID = (
    mozilla => '{86c18b42-e466-45a9-ae7a-9b95ba6f5640}',
    firefox => '{ec8030f7-c20a-464f-9b0e-13a3a9e97384}',
    flock => '{a463f10c-3994-11da-9945-000d60ca027b}',
    seamonkey => '{92650c4d-4b8e-4d2a-b7eb-24ecf4f6b63a}',
    thunderbird => '{3550f703-e582-4d05-9a08-453d09bdfdc6}',
    nvu => '{136c295a-4a5a-41cf-bf24-5cee526720d5}',
    sunbird => '{718e30fb-e89b-41dd-9da7-e25a45638b28}',
    netscape => '{3db10fab-e461-4c80-8b97-957ad5f8ea47}',
);

# for install.rdf generation:
template main => sub {
    my ($self, $xpifile) = @_;
    xml_decl { 'xml', version => '1.0', encoding => 'UTF-8' };
    RDF {
        attr {
            'xmlns' => "http://www.w3.org/1999/02/22-rdf-syntax-ns#",
            'xmlns:em' => 'http://www.mozilla.org/2004/em-rdf#'
        }
        Description {
            attr { about => 'urn:mozilla:install-manifest' }
            em::id { $xpifile->id }
            em::name { $xpifile->name }
            em::description { $xpifile->description }
            em::version { $xpifile->version }
            em::creator { $xpifile->creator };
            for (@{ $xpifile->developers }) {
                em::developer { $_ }
            }
            for (@{ $xpifile->contributors }) {
                em::contributor { $_ }
            }

            my $targets = $xpifile->targets;
            while (my ($app, $ver_range) = each %$targets) {
                outs_raw("\n\n  <!-- $app, version $ver_range->[0] - $ver_range->[1] -->");
		em::targetApplication {
                    Description {
                        em::id { get_uuid($app) }
                        em::minVersion { $ver_range->[0] }
                        em::maxVersion { $ver_range->[1] }
                    }
                }
            }
            em::homepageURL { $xpifile->homepageURL };
            if ($xpifile->updateURL) {
                em::updateURL { $xpifile->updateURL }
            }
            if ($xpifile->iconURL) {
                em::iconURL { $xpifile->iconURL }
            }

        }
    }
};

sub get_uuid {
    my $name = shift;
    my $key = lc($name);
    #warn $key;
    ### %UUID
    my $value = $UUID{$key};
    if (!$value) {
        die "Can't find UUID for target app $name\n";
    }
    return $value;
}

1;



lib/XUL/App.pm
package XUL::App;

use strict;
use warnings;
use base qw/ Class::Data::Inheritable /;

our $VERSION = '0.07';

__PACKAGE__->mk_classdata('FILES' => {});

our ($ID, $APP_NAME);

1;
__END__

=head1 NAME

XUL::App - Nifty XUL apps in a XUL::App

=head1 VERSION

This document describes XUL::App 0.07 released on July 11, 2008.

=head1 SYNOPSIS

    $ xulapp app --name YSearchAll

    $ cd YSearchAll

    $ xulapp view --name Overlay --type overlay

    # Edit lib/YSearchAll/App.pm to add the following lines:
    package YSearchAll::App;
    our $VERSION = '0.07';

    use XUL::App::Schema;
    use XUL::App schema {
    +     xulfile 'overlay.xul' =>
    +         generated from 'YSearchAll::View::Overlay',
    +         includes qw( xulapp/jquery.js overlay.js ),
    +         overlays 'chrome://browser/content/browser.xul';
    +
        xpifile 'ysearchall.xpi' =>
            name is 'YSearchAll',
            id is 'ysearchall@agentz.agentz-office',
            version is '0.0.1',
            targets {
                Firefox => ['2.0' => '3.0.*'],  # FIXME
                Mozilla => ['1.5' => '1.8'],  # FIXME
            },
            creator is 'The YSearchAll development team',
            developers are ['agentz'],
            contributors are [],
            homepageURL is 'http://searchall.agentz.org', # FIXME
            iconURL is '',  # like 'chrome://helloworld/content/logo.png'
            updateURL is ''; # This should not set for AMO extensions.
    };
    1;

    # Add and edit js/ysearchall.js manually
    $ xulapp overlay -p dev11
    $ xulapp bundle .  # generate XPI ready for deployment

=head1 DESCRIPTION

XUL::App is a nifty Firefox extension development framework based on Perl. It has a lot of parallels with L<Jifty>. In particular, this framework allows you to build real-world modern Firefox extensions using Perl. But the resulting XPI installation file is completely portable and contains I<0> Perl.

This framework has the following highlights:

=over

=item *

L<Jifty> love on the Firefox extension development land.

=item *

Building a realworld Firefox extension can be much easier than GreaseMonkey hacks.

=item *

Use 100% Perl to specify XUL layout via L<Template::Declare>, no XML pain.

=item *

Automatic extension registration/unregistration for easy debugging. And no more frustration from Firefox's XUL/ext cache. (XUL::App will ensure the developer version defeats Firefox's cache.)

=item *

Transparent C<jar>-based XPI bundling, no F<chrome.manifest>, L<install.rdf>, and zip panic. XUL::App generates all of them for you according to your (declarative) Perl code.

=item *

The resulting XPI file contains B<0%> Perl and could run everywhere (Win32, Mac, Linux, and etc). It's a compiler-style framework.

=item *

I18N support via the L<Locale::Maketext::Lexicon> module (same as L<Jifty>, but
actually emulated by XUL's native I18N mechanism).

=back

Currently the module is still in B<alpha> stage and we're in severe lack of documentation and tests. (Although a real-world fully-fledged Firefox extension named SearchAll (L<https://addons.mozilla.org/en-US/firefox/addon/5712>) is already developed atop it.

You can get a lot of information from the slides that I used in the talk:

L<http://agentzh.org/misc/slides/xulapp/xulapp.xul>  (a JS-enabled Firefox is required to view these slides)

If you're having problems in displaying the slides given above, please try out the PowerPoint (PPT) version below:

L<http://agentzh.org/misc/slides/xulapp.ppt>

or the PDF version:

L<http://agentzh.org/misc/slides/xulapp.pdf>

=head1 CAVEATS

=over

=item *

Because XUL::App wites to your ~/.mozilla/firefox directory directly, it's always recommended to backup that directory to somewhere else to prevent unexpected corruption occurring to your Firefox profiles.

=item *

After applying Firefox updates or using a new Firefox profile, it's required to restart Firefox one more time to load your extensions. So please don't fire bug report regarding that.

=item *

Only Firefox 2.0.0.* and Firefox 3.0 on Linux has been tested against this framework. So you're warned when you're going to do XUL::App hacking on other platforms like Windows.

(BTW, I'm very willing to apply patches for other platforms.)

=back

=head1 SOURCE CONTROL

You can always get the latest source of XUL::App from the following SVN repository:

L<http://svn.openfoundry.org/xulapp/trunk/>

There's very little POD docs, please read README and the slides for details.

I really need help in improving this module's docs, tests, and implementation. If you find this thing useful and feel like contributing to it, please write to me and get a commit bit! ;)

=head1 SAMPLES

=over

=item *

SearchAll is a real-world Firefox extension that is built upon XUL::App, which can serve as a big demo for the usage of the framework:

L<http://svn.openfoundry.org/searchall/trunk/>

And it's already on the mozilla official site AMO:

L<https://addons.mozilla.org/en-US/firefox/addon/5712>

=item *

ExportReader: A Firefox extension to dump entries from Google Reader to JSON:

L<http://svn.openfoundry.org/xulapp/trunk/demo/ExportReader/>

This extension only costed me about 10 lines of Perl and 20 lines of JavaScript, in addition to a few shell commands. The process is easy and enjoyable.

Because it's mostly for personal use, The usage of this addon deserves some explanation though:

=over

=item 1.

Click the XPI file from your Firefox browser:

L<http://svn.openfoundry.org/xulapp/trunk/demo/ExportReader/exportreader.xpi>

After installation, please remember to restart to browser.

=item 2.

Enter the following chrome URL in your Firefox's address bar:

L<chrome://exportreader/content/exportreader.xul>

Then you will see the main UI of this extension.

=item 3.

Login to your Google Reader and then click the site on the left-hand-side menu whose entries are what you want to export, say, "chromatic's Journal". Ensure that the "Expanded View" is used in the right-hand-side on the Google Reader page.

=item 4.

Click the "Extract!" button on the right-upper corner. Then you'll see an alert dialog saying how many entries were found. Click "OK" then the corresponding JSON string will be in the right-most textbox which is ready for copy-and-paste.

Note that, by defaul, Google Reader lazily loads just the top 5 entries or so. So in order to export all the entries in the subscribed site, say, "chromatic's Journal", you'll have to scroll down the Reader's Expanded View and force it to retrieve more entries.

=back

=item *

A helloworld sample extension is given in my XUL::App talk's slides:

L<http://agentzh.org/misc/slides/xulapp.pdf>

=back

=head1 INSTALLATION

    perl Makefile.PL
    make
    sudo make install

=head1 BUGS

Sadly XUL::App does not run on Win32 yet. I've only tested it on Ubuntu Linux so far. If you have any problems or would love to help, please let me know ;)

=head1 AUTHOR

Agent Zhang <agentzh@yahoo.cn>

=head1 COPYRIGHT AND LICENSE

Copyright 2007, 2008 by Yahoo! China EEEE Works, Alibaba Inc. (L<http://www.eeeeworks.org>)

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Artistic or GPL.

=head1 SEE ALSO

L<xulapp>, L<Locale::Maketext::Lexicon>, L<Template::Declare>, L<Jifty>.

